### ISSUE TITLE
mIoU for segmentation

### ISSUE THREAD

**Issue by EthanAbitbol3** [ultralytics/ultralytics#3456]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello and thank you for your work! I am currently testing segmentation models and I'm trying to add an mIoU (mean Intersection over Union) function to the metric.py file so that I can compare it with other state-of-the-art semantic segmentation models. However, I'm facing some difficulties, and I was wondering if anyone has any suggestions on how to do it?

Thank you and have a great day!

**Comment by glenn-jocher** [comment]
@EthanAbitbol3 hi there! Thank you for your interest in YOLOv8 and for reaching out with your question. We appreciate your support!

Adding an mIoU (mean Intersection over Union) function to the `metric.py` file in YOLOv8 for comparing segmentation models sounds like a great idea. However, YOLOv8 is primarily designed for object detection rather than semantic segmentation. 

That being said, if you want to implement mIoU in YOLOv8, you would need to modify the codebase accordingly. Here's a brief overview of the steps you can take to add the mIoU function:

1. Start by importing any necessary libraries or functions for calculating IoU, such as `numpy`.
2. In the `metric.py` file, define a new function called `calculate_mIoU` or something similar.
3. Within this function, calculate the Intersection over Union (IoU) for each predicted segmentation mask and the corresponding ground truth mask.
4. Compute the mIoU by taking the average of the IoU values across all the masks.
5. You may also need to update other parts of the code to accommodate the mIoU calculation, such as modifying the output format or storing the mIoU values for later comparison.

Please note that adding mIoU to YOLOv8 might require significant modifications to the existing codebase. It's important to carefully test and validate your changes to ensure the results are accurate and reliable.

If you're looking for state-of-the-art semantic segmentation models, we would recommend exploring dedicated frameworks or repositories that focus specifically on semantic segmentation tasks.

I hope this information helps! If you have any further questions or need additional assistance, please feel free to ask. Have a great day too!

**Comment by glenn-jocher** [comment]
@EthanAbitbol3 hi there! Thank you for your effort in modifying the `metric.py` file in YOLOv8 and for sharing your approach.

Regarding the issue you mentioned, it seems like you attempted to calculate the average IoU per image by modifying the `_process_batch` function in the `val.py` file. You then stored these averaged IoU values in a list and took the mean of that list in the `update_metrics` function.

However, you mentioned that the calculated IoU values are extremely poor, despite the mean Average Precision (mAP) being correct. It's important to note that mAP and IoU are different evaluation metrics. While mAP provides an overall evaluation of object detection performance, IoU specifically measures the overlap between predicted and ground truth bounding boxes or masks.

It's possible that there might be a discrepancy in how the IoU values are being calculated or stored, leading to poor results. I would recommend reviewing the modifications you made to ensure that the IoU calculation is accurate and that the values are being properly stored and averaged.

Here are a few suggestions for troubleshooting this issue:

1. Double-check your modifications in the `_process_batch` function to ensure that the average IoU per image is being calculated correctly. Verify that the correct values are being returned by the `mask_iou` function.
2. Review the modifications in the `update_metrics` function to confirm that the IoU values are being stored and averaged correctly. Check the implementation of storing the values in the list and calculating the mean.
3. Consider analyzing a smaller subset of your dataset to validate the correctness of your modifications. This can help identify any issues in the calculations or data processing.
4. Ensure that the ground truth masks and predicted masks are correctly aligned and in the expected format for calculating IoU. Any discrepancies here may lead to inaccurate results.

By carefully reviewing and validating your modifications, you should be able to identify any issues that may be causing the poor IoU values. Don't hesitate to reach out if you need further assistance or have any additional questions. Good luck with your work on YOLOv8!

**Comment by bit-scientist** [comment]
Hi, @EthanAbitbol3, did you get what you were trying to accomplish? May I get your help on that?

**Comment by SmallWhiteGZ** [comment]
> Thank you very much for your response. I started to modify the metric.py file, and I noticed that there is already a function called mask_iou, which, if I understood correctly, returns the IoU for each instance of each image. So I tried to return the average per image like this in the _process_batch function of the val.py file (line 160):
> 
> ![image](https://user-images.githubusercontent.com/43862103/250494143-9d929339-1235-4b82-8043-2a23562d638b.png)
> 
> Following this, I modified the update_metrics function in val.py to store all the averaged IoU values in a list (line 107) and then calculate the mean of this list. ![image](https://user-images.githubusercontent.com/43862103/250495349-c510bc91-7b10-4740-9cc9-d58d50184b26.png)
> 
> The problem is that the values are extremely poor, despite the mAP being correct. For example:
> 
> ![image](https://user-images.githubusercontent.com/43862103/250496638-f769b018-3856-4cfd-89db-a6dd007d478a.png)

the shape of iou is (num_gt,num_pre), you should add iou.max(dim=1) to get the correct iou.

**Comment by bit-scientist** [comment]
@EthanAbitbol3 share your solution after you think you got it right. We can benefit from that, thank you!

**Comment by glenn-jocher** [comment]
@SmallWhiteGZ We appreciate your efforts in working towards the resolution of this problem and for bringing us along your journey. Once you've designed a successful implementation and you're confident about its correctness, your insights would be welcome. Your contributions have the potential to help shape and enhance the YOLOv8 community. Thank you for your ongoing support and active participation!

**Comment by glenn-jocher** [comment]
@manhcuong02 hey there! 🌟 It's fantastic to hear about your exploration into segmentation metrics and your success in integrating mIoU into YOLOv8 for your project. Your initiative to dive into the source code and experiment is truly commendable. 

Regarding your question about contributing your code, we're always excited to see community contributions that can enhance YOLOv8. To get started with contributing, you can follow these steps:
1. Fork the YOLOv8 repository on GitHub to create a copy under your account.
2. Clone your forked repository to your local machine and make your changes.
3. Push your changes to your forked repository.
4. Create a pull request from your repository back to the original YOLOv8 repository.

Before you submit your pull request, please ensure your code is well-documented and tested across different scenarios to ensure robustness. Given the sensitivity of your project, make sure no confidential information is included in your contribution.

If you're new to contributing, GitHub has some great resources to guide you through the process. And of course, if you have any questions or need further assistance, feel free to reach out. We're here to help!

Thanks for your kind words and for considering contributing to the YOLOv8 community. Your efforts not only help improve the project but also inspire others in the AI research community. Looking forward to seeing your contribution! 🚀

**Comment by manhcuong02** [comment]
@glenn-jocher Thank you for your response. I will experiment with them on various datasets with a small number of epochs. If the source code continues to run smoothly, I will update you in the near future and will create a pull request from my repository back to the original YOLOv8 repository. I sincerely appreciate your assistance.

**Comment by manhcuong02** [comment]
@glenn-jocher I have a query regarding the output of your segmentation model. Its output exceeds the number of classes I specified in the YAML file. For instance, in my scenario, I set the number of classes to 2; however, the pred_masks output contains more than 2 classes such as 4, 5, 8. I found this in the segment/val.py file by printing it as shown in the image below:
![image](https://github.com/ultralytics/ultralytics/assets/77909327/a252f34e-3e68-4777-ada7-05457e66a5d4)

And the result I obtained is:
![image](https://github.com/ultralytics/ultralytics/assets/77909327/155a1ac1-d90e-497e-b9a1-2092e4e701f4)
Please let me know if I am misunderstanding the output of your segmentation model. I appreciate your assistance greatly.

**Comment by glenn-jocher** [comment]
@manhcuong02 hey there! 😊 It sounds like you're encountering an interesting scenario with the segmentation model outputs. From what you've described, it seems like there might be a discrepancy in how the model interprets class labels during training and validation phases.

One possible explanation for seeing unexpected class labels in `pred_masks` could be related to how the model's predictions are being processed or a mismatch in the dataset labels. It's also worth checking if there's any data augmentation or preprocessing step that might inadvertently alter class labels during the training phase but not during validation.

For a quick check, you could ensure that your dataset's labeling is consistent and that the class indices in your YAML file match those used during training and validation. Here's a small snippet to verify the unique classes present in your `pred_masks`:

```python
unique_classes = pred_masks.unique()
print(f'Unique classes in pred_masks: {unique_classes}')
```

If the issue persists and you're confident it's not related to the dataset or preprocessing steps, it might be worth exploring further into the model's training logic to see if there's any point where class predictions could be misinterpreted.

Thanks for bringing this to our attention, and we appreciate your patience as we work through this together. If you have any more insights or updates based on these suggestions, feel free to share!

**Comment by manhcuong02** [comment]
@glenn-jocher From my understanding, each channel in pred_masks has values of 0 and 1, so using pred_masks.unique() to check the number of classes in the dataset is not feasible because it will return 2 values, 0 and 1, regardless of the number of classes. The number of channels in pred_masks corresponds to the number of classes returned by the segmentation model. I also encountered this issue when using model.val() in some cases. I'm not sure what the reason is. Thank you for your response.

**Comment by glenn-jocher** [comment]
@manhcuong02 No worries at all! 😊 It's great to hear that you've had a breakthrough in understanding the segmentation model's input and output data. Indeed, the `pred_masks` channels correspond to the model's class predictions, with each channel representing a class and containing binary values for the presence of that class in each pixel.

It sounds like you're on the right track now, and diving deeper into the source code is a fantastic way to get there. If you have any further questions or if there's anything specific you'd like to discuss regarding the implementation of the mIoU metric or any other aspect, feel free to reach out. Your initiative to enhance the model with additional metrics is highly valued, and I'm looking forward to seeing your updates. Keep up the great work, and thank you for contributing to the YOLOv8 community! 🚀

**Comment by manhcuong02** [comment]
@glenn-jocher Although I am facing challenges due to the differences between ground truth and predicted masks, I have found a solution to it. This may be difficult for me to implement and may take time to address. However, for now, I hope that at least it can run correctly. Then, I will focus on improving the speed afterward. Thank you very much!

**Comment by manhcuong02** [comment]
@glenn-jocher Hi there! I need your assistance. First, I want to say that I am almost done with calculating mIoU for the problem I am working on. I have understood and addressed the issues as mentioned above. To verify this, I have displayed the number of predicted instances during the validation process. (However, the number of predicted instances may be larger or smaller than the actual number of instances depending on the value of the `confidence parameter` we set. During training, the default `confidence` is set very small, at `0.0001`, so the number of predicted instances may be much larger than the actual number.) The image below shows the result of calculating mIoU for the problem I am currently dealing with:
![segment](https://github.com/ultralytics/ultralytics/assets/77909327/f392f30c-8bc1-49b2-b553-1e6f5b2f592d)

Similarly, to avoid disrupting other tasks within YOLOv8, I also trained a basic object detection model for validation purposes. It is reassuring that everything is still running smoothly, and the results are displayed in the image below:
![detect](https://github.com/ultralytics/ultralytics/assets/77909327/cfe6e91b-7674-492b-ac14-84a11d990e24)

To validate my modifications in the mIoU calculation process, I need to apply it to a more general problem. Therefore, I also used the `coco128-seg` dataset to validate my modified source code. While there were some unexpected errors encountered when working with multi-classes, everything has been resolved satisfactorily. **However, I cannot verify the mIoU results on this dataset. I would appreciate your input on this matter**. Thank you. Below is the image showing the results of calculating mIoU on my `coco128-seg` dataset:
![coco128seg](https://github.com/ultralytics/ultralytics/assets/77909327/b60124ea-0424-4e51-a42d-f9104cb0cb81)
![coco128seg-2](https://github.com/ultralytics/ultralytics/assets/77909327/dd8c7409-f14a-4ccf-8b52-cca257964639)

Thank you so much!!

**Comment by manhcuong02** [comment]
@glenn-jocher 
Thank you for your suggestions. I will try with other segmentation models to verify my mIoU calculation results. However, I have a question about how to determine the acceptable error between two methods. Additionally, I would like to inquire about how to handle the scenario where a predicted segmentation result in an image corresponds to a class not present in the ground truth labels. Should I ignore this case? I sincerely appreciate your assistance.

**Comment by glenn-jocher** [comment]
@manhcuong02 You're welcome! 😊 For verifying your mIoU calculations, a small error margin, typically within a few percentage points, is generally acceptable due to differences in implementation details. As for predicted classes not present in the ground truth, it's standard to consider these predictions as false positives in your evaluation metrics. This means they would negatively impact precision but are necessary to include for a comprehensive assessment of your model's performance. Keep up the great work, and don't hesitate to reach out if you have more questions!

**Comment by manhcuong02** [comment]
@glenn-jocher Thank you for your response. Could you provide me with some suggestions on how to handle false positive cases? I am currently unsure how to address this issue. Thank you very much.

**Comment by glenn-jocher** [comment]
@manhcuong02 Absolutely, happy to help! 😊 For handling false positives, your approach is on the right track. When a predicted instance doesn't match any ground truth label (class mismatch), you can indeed ignore its IoU value. However, instead of setting IoU for the unmatched class to 0, consider not altering the IoU sum for that class but still incrementing the count. This way, you accurately reflect the absence of a correct prediction for that class, impacting the mIoU calculation appropriately. Here's a quick snippet:

```python
if predicted_class != ground_truth_class:
    # Don't add to IoU sum, but increment count for ground truth class
    count[ground_truth_class] += 1
```

This method ensures that your mIoU metric remains a fair representation of model performance across all classes. Keep up the great work!

**Comment by manhcuong02** [comment]
@glenn-jocher Hello there. Currently, I have managed to edit the mIoU calculation after excluding false positive predicted instances and have also optimized my edits to be more concise. As you mentioned earlier, comparing my mIoU values with those of another segmentation model could be a good approach. However, I haven't been able to find any pre-trained segmentation model on the COCO dataset to do this, and I also lack the resources to accomplish that. Yet, when examining the values of Recall, mAP50, and mAP50-95, I noticed that my mIoU values are approximately similar. If the values of these metrics are low for a certain class, the mIoU for that class tends to be low as well. I believe this may indicate that my approach is correct. I'm not sure if my thinking is accurate, and I would appreciate your insights on this matter. Here are the results of my adjustments on the coco128-seg dataset:
![image](https://github.com/ultralytics/ultralytics/assets/77909327/1cff1c58-8d0b-4a7c-a745-efe0b6574124)
![image](https://github.com/ultralytics/ultralytics/assets/77909327/818b29a9-bf31-4404-b6f8-19217fb2ec16)
Thank you so much!

**Comment by glenn-jocher** [comment]
@manhcuong02 Hello! 😊 Great job on refining your mIoU calculation and making your code more concise. It's quite insightful to compare mIoU with other metrics like Recall, mAP50, and mAP50-95. If you're seeing a correlation where classes with lower values in these metrics also show lower mIoU, it does suggest your approach is aligning well with general model performance indicators. This correlation is expected as all these metrics aim to measure the accuracy of your model, albeit from slightly different perspectives.

Without direct comparison to another model, this method of cross-referencing metrics can indeed serve as a good sanity check for your mIoU implementation. Keep up the fantastic work, and remember, the journey of refining and understanding your model's performance is ongoing. If you have more questions or need further assistance, feel free to reach out! 🚀

**Comment by manhcuong02** [comment]
@glenn-jocher Before delving into the details of variables in your segmentation model, I wrote additional auxiliary functions for my mIoU calculation process. However, later on, they were removed, and I used your existing functions. This may demonstrate that the model's performance regarding speed will not be affected because I only added a few simple calculations on the matrix. As for my modification, currently, I have not been able to detect any further issues to improve upon. I hope to be able to submit a pull request to your repository so that you and other users can identify and address this issue. Thank you sincerely.

**Comment by melikycode** [comment]
@manhcuong02 it is awesome work you are doing. Were you able to add it ?

**Comment by manhcuong02** [comment]
@melikycode Yes, I am submitting a pull request to the author of the repository. However, the `tests check` process for `YOLOv8` is encountering some issues while validating with the `FastSAM model`. Therefore, my pull request cannot be merged with the author's original repository yet. We are discussing this issue. You can find more details here: [link to the pull request](https://github.com/ultralytics/ultralytics/pull/9109)

**Comment by zai1318** [comment]
> riment with them on various datasets with a small number of epochs

i dont understand this , in one image i can see the miou values but then your next image does not show any miou values and yet you say they are being implemented correctly ... i implemented the same changes but i am unable to make it show the miou results, can you please help me with that . 
Thank you

**Comment by manhcuong02** [comment]
@zai1318 As I understand, you're referring to the two images in [this comment](https://github.com/ultralytics/ultralytics/issues/3456#issuecomment-1969356722). The first image shows the IOU because I implemented it only for the segmentation task, while the second image is from my experiment on the detection task to see if my changes to segmentation would have any impact on other existing tasks. Thank you for your interest!

**Comment by glenn-jocher** [comment]
Thank you for your update. It sounds like you're making progress, and focusing on correctness first is a great approach. If you have any further questions or need assistance, feel free to ask.

**Comment by zai1318** [comment]
> @zai1318 As I understand, you're referring to the two images in [this comment](https://github.com/ultralytics/ultralytics/issues/3456#issuecomment-1969356722). The first image shows the IOU because I implemented it only for the segmentation task, while the second image is from my experiment on the detection task to see if my changes to segmentation would have any impact on other existing tasks. Thank you for your interest!

yes exactly i applied it the same way on the segmentation task but i did not got any results like that, this is why i am confused that how to make miou values appear after training the model on segmentation task. Thank you

**Comment by glenn-jocher** [comment]
@manhcuong02 it's great to hear you've found a solution. If you need further assistance or have questions as you work on improving the speed, feel free to reach out. Good luck!

**Comment by manhcuong17072002** [comment]
@zai1318 @glenn-jocher This issue was resolved quite a while ago and is currently no longer maintained. Currently, Ultralytics has released many updates, and Ultralytics Assistant has pushed numerous commits to the repo I forked from Ultralytics. This may cause some issues that make my code no longer function as before. If I have some free time, I will review this issue again. Thank you

**Comment by manhcuong02** [comment]
@zai1318 @glenn-jocher This issue was resolved quite a while ago and is currently no longer maintained. Currently, Ultralytics has released many updates, and Ultralytics Assistant has pushed numerous commits to the repo I forked from Ultralytics. This may cause some issues that make my code no longer function as before. If I have some free time, I will review this issue again. Thank you

**Comment by glenn-jocher** [comment]
Great to hear you've found a solution! If you need further assistance or have any questions as you implement it, feel free to reach out. Good luck!

**Comment by glenn-jocher** [comment]
It's great to hear that you've found a solution to the issue with the mask structures. If you need further assistance or have any questions as you work on improving the speed, feel free to reach out. Good luck with your implementation!

**Comment by manhcuong02** [comment]
@zai1318 By the way, someone recently asked me about mIoU in segmentation tasks. I double-checked and tried running a command to validate the existing COCO dataset and it returned reasonable mIoU values for each class. You could try and review your results. Thanks.

![image](https://github.com/user-attachments/assets/57f886a9-125f-4896-93b3-8557c99db068)
